jctools-core/src/main/java/org/jctools/queues/MpscProgressiveChunkedQueue.java

```diff
@@ -0,0 +1,553 @@
+        fieldOffset(MpscProgressiveChunkedQueueConsumerFields.class, "consumerIndex");
+
+    private volatile long consumerIndex;
+    protected AtomicChunk<E> consumerBuffer;
+
+    @Override
+    public final long lvConsumerIndex()
+    {
+        return consumerIndex;
+    }
+
+    final long lpConsumerIndex()
+    {
+        return UNSAFE.getLong(this, C_INDEX_OFFSET);
+    }
+
+    final void soConsumerIndex(long newValue)
+    {
+        UNSAFE.putOrderedLong(this, C_INDEX_OFFSET, newValue);
+    }
+}
+
+abstract class MpscProgressiveChunkedQueuePad4<E> extends MpscProgressiveChunkedQueueConsumerFields<E>
+{
+    long p0, p1, p2, p3, p4, p5, p6, p7;
+    long p10, p11, p12, p13, p14, p15, p16, p17;
+}
+
+// $gen:ordered-fields
+abstract class MpscProgressiveChunkedQueueFreeBuffer<E> extends MpscProgressiveChunkedQueuePad4<E>
+{
+    private static final long F_BUFFER_OFFSET =
+        fieldOffset(MpscProgressiveChunkedQueueFreeBuffer.class, "freeBuffer");
+
+    private volatile AtomicChunk<E> freeBuffer;
+
+    final AtomicChunk<E> lpFreeBuffer()
+    {
+        return (AtomicChunk<E>) UNSAFE.getObject(this, F_BUFFER_OFFSET);
+    }
+
+    final AtomicChunk<E> lvFreeBuffer()
+    {
+        return this.freeBuffer;
+    }
+
+    final void soFreeBuffer(AtomicChunk<E> buffer)
+    {
+        UNSAFE.putOrderedObject(this, F_BUFFER_OFFSET, buffer);
+    }
+
+    final boolean casFreeBuffer(AtomicChunk<E> expected, AtomicChunk<E> newValue)
+    {
+        return UNSAFE.compareAndSwapObject(this, F_BUFFER_OFFSET, expected, newValue);
+    }
+}
+
+abstract class MpscProgressiveChunkedQueuePad5<E> extends MpscProgressiveChunkedQueueFreeBuffer<E>
+{
+    long p0, p1, p2, p3, p4, p5, p6, p7;
+    long p10, p11, p12, p13, p14, p15, p16, p17;
+}
+
+/**
+ * An MPSC array queue which starts at <i>initialCapacity</i> and grows unbounded in linked chunks.<br>
+ * Differently from {@link MpscUnboundedArrayQueue} it is designed to provide a better scaling when more
+ * producers are concurrently offering.
+ *
+ * @param <E>
+ * @author https://github.com/franz1981
+ */
+public class MpscProgressiveChunkedQueue<E> extends MpscProgressiveChunkedQueuePad5<E>
+    implements MessagePassingQueue<E>, QueueProgressIndicators
+{
+    private final int chunkMask;
+    private final int chunkShift;
+
+    public MpscProgressiveChunkedQueue(int chunkSize)
+    {
+        chunkSize = Pow2.roundToPowerOfTwo(chunkSize);
+        final AtomicChunk<E> first = new AtomicChunk(0, null, chunkSize);
+        soProducerBuffer(first);
+        consumerBuffer = first;
+        chunkMask = chunkSize - 1;
+        chunkShift = Integer.numberOfTrailingZeros(chunkSize);
+        soFreeBuffer(new AtomicChunk(-1, null, chunkSize));
+    }
+
+    private AtomicChunk<E> producerBufferOf(AtomicChunk<E> producerBuffer, long chunkIndex)
+    {
+        long jumpBackward = producerBuffer == null ? -1 : producerBuffer.index - chunkIndex;
+        while (jumpBackward < 0)
```

Comment:
You can optimize for common case here where the index matches (jumpBackward == 0)

Reply:
So, perform the same check outside this slow path method to avoid it to be called?